src/cdomains/c2poDomain.ml

```diff
@@ -0,0 +1,1337 @@
+      (* If they are in the same equivalence class and they overlap, then they are equal *)
+      (if T.equal q' q && there_is_an_overlap s s' diff then true
+       else
+         (* If we have a disequality, then they are not equal *)
+       if neq_query (Some cc) (t,v,Z.(z'-z)) then false else
+         (* or if we know that they are not equal according to the query MayPointTo*)
+       if GobConfig.get_bool "ana.c2po.askbase" then (may_point_to_same_address ask t v Z.(z' - z) cc)
+       else true)
+      || (may_be_equal ask cc s t1 v)
+    | Deref _, _ -> false (* The value of addresses or auxiliaries never change when we overwrite the memory*)
+    | Addr _ , _ | Aux _, _ -> T.is_subterm t1 t2
+
+  (**Returns true iff by assigning to t1, the value of t2 could change.
+     The parameter s is the size in bits of the variable t1 we are assigning to. *)
+  let may_be_equal ask cc s t1 t2 =
+    match cc with
+    | None -> false
+    | Some cc ->
+      let res = (may_be_equal ask cc s t1 t2) in
+      if M.tracing then M.tracel "c2po-maypointto" "MAY BE EQUAL: %s %s: %b\n" (T.show t1) (T.show t2) res;
+      res
+
+  let rec may_point_to_one_of_these_adresses ask adresses cc t2 =
+    match t2 with
+    | Deref (v, z',_) ->
+      (may_point_to_address ask adresses v z' cc)
+      || (may_point_to_one_of_these_adresses ask adresses cc v)
+    | Addr _ | Aux _ -> false
+
+end
+
+module D = struct
+
+  include Printable.StdLeaf
+
+  type domain = t option [@@deriving ord, hash]
+  type t = domain [@@deriving ord, hash]
+
+  (** Convert to string *)
+  let show x = match x with
+    | None -> "⊥\n"
+    | Some x -> show_conj (get_normal_form x)
+
+  let show_all = function
+    | None -> "⊥\n"
+    | Some x -> show_all x
+
+  include Printable.SimpleShow(struct type t = domain let show = show end)
+
+  let name () = "c2po"
+
+  let equal x y =
+    if x == y then
+      true
+    else
+      let res =
+        match x,y with
+        | None, None -> true
+        | Some cc1, Some cc2 ->
+          if cc1 == cc2 then
+            true
+          else
+            (* add all terms to both elements *)
+            let terms = SSet.union (SSet.union cc1.set (BlDis.term_set cc1.bldis))
+                (SSet.union cc2.set (BlDis.term_set cc2.bldis)) in
+            let cc1, cc2 = Option.get (insert_set (Some cc1) terms), Option.get (insert_set (Some cc2) terms) in
+            equal_eq_classes cc1 cc2
+            && equal_diseqs cc1 cc2
+            && equal_bldis cc1 cc2
+        | _ -> false
+      in if M.tracing then M.trace "c2po-equal" "equal. %b\nx=\n%s\ny=\n%s" res (show_all x) (show_all y);res
+
+  let empty () = Some {uf = TUF.empty; set = SSet.empty; map = LMap.empty; diseq = Disequalities.empty; bldis = BlDis.empty}
+
+  let init () = init_congruence []
+
+  let bot () = None
+  let is_bot x = Option.is_none x
+  let top () = empty ()
+  let is_top = function None -> false
+                      | Some cc -> TUF.is_empty cc.uf
```

Comment:
Is this correct? Can I not have, e.g., a (block) disequality but still an empty TUF?